### PR TITLE
Update token removal range to keep leading indentation

### LINKED ISSUE #1039: Token cleanup removes line indentation
The new whitespace token cleanup introduced in #1028 is a bit too aggressive. It will remove line indentation in certain situations. For example:

```
    foo.
```

Saying `"chuck fine"` will remove the line indentation, leaving the `.` at the start of the line:

```
.
```

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Fixes #1039

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet_html)

### CODE REVIEW

File: src/processTargets/targetUtil/insertionRemovalBehaviors/DelimitedSequenceInsertionRemovalBehavior.ts [hunk 1]
```diff
@@ -5,15 +5,10 @@ import { Target } from "../../../typings/target.types";
  * Constructs a removal range for the given target that includes either the
  * trailing or leading delimiter
  * @param target The target to get the removal range for
- * @param contentRange Can be used to override the content range instead of
- * using the one on the target
  * @returns The removal range for the given target
  */
-export function getDelimitedSequenceRemovalRange(
-  target: Target,
-  contentRange?: Range
```
Comment (pokey) [inline src/processTargets/targetUtil/insertionRemovalBehaviors/DelimitedSequenceInsertionRemovalBehavior.ts:14]: This parameter was confusing, and only used in one place, and removing it didn't break any tests, so I removed it.  Fwiw the place where we were using it was actually broken anyway 😅

File: src/processTargets/targetUtil/insertionRemovalBehaviors/TokenInsertionRemovalBehavior.ts [hunk 2]
```diff
@@ -59,45 +58,81 @@ export function getTokenTrailingDelimiterTarget(
...
-      document,
-      fullText,
-      fullTextOffset,
-      removalRange
-    );
-
-    // Using removal range has not merged any tokens. Removal range is ok to use.
-    if (numTokensContentRangeRemoved === numTokensRemovalRangeRemoved) {
-      return removalRange;
+export function getTokenRemovalRange(target: Target): Range {
```
Comment (pokey) [inline src/processTargets/targetUtil/insertionRemovalBehaviors/TokenInsertionRemovalBehavior.ts:63]: I decided to rework this function, because I think that `getDelimitedSequenceRemovalRange` was obscuring the full picture here, and we were really only relying on it for the convention of preferring trailing over leading.  We do lose the fact that we have a single source of truth for that convention, but given how special this token removal code has become I don't think it was worth working around that function

File: src/processTargets/targetUtil/insertionRemovalBehaviors/TokenInsertionRemovalBehavior.ts [hunk 3]
```diff
@@ -59,45 +58,81 @@ export function getTokenTrailingDelimiterTarget(
...
+    target.getTrailingDelimiterTarget()?.contentRange ?? makeEmptyRange(end);
+
+  const fullLineRange = expandToFullLine(editor, contentRange);
+
+  if (
+    leadingWhitespaceRange.union(trailingWhitespaceRange).isEqual(fullLineRange)
+  ) {
+    // If we would just be leaving a line with whitespace on it, we delete the
+    // whitespace
+    return fullLineRange;
```
Comment (pokey) [inline src/processTargets/targetUtil/insertionRemovalBehaviors/TokenInsertionRemovalBehavior.ts:80]: This seemed like nice behaviour, so I added it while I was here.  Note that we almost got this behaviour with your version, but not if there was trailing whitespace on the line.  Have now added test for that

File: src/processTargets/targets/ScopeTypeTarget.ts [hunk 4]
```diff
@@ -90,7 +90,7 @@ export default class ScopeTypeTarget extends BaseTarget {
 
   getRemovalRange(): Range {
     return this.removalRange_ != null
-      ? getTokenRemovalRange(this, this.removalRange_)
+      ? this.removalRange_
```
Comment (pokey) [inline src/processTargets/targets/ScopeTypeTarget.ts:93]: This was trying to cleanup whitespace on removal range passed in, but it wasn't doing it correctly, and we don't do that when they pass in delimiters, so it seemed inconsistent.  We can revisit in the future if we want to clean up whitespace on removal ranges.  As mentioned above, nothing failed when I changed this

File: src/test/suite/fixtures/recorded/scopes/token/ditchFine.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  spokenForm: ditch fine
+  version: 3
+  targets:
+    - type: primitive
+      mark: {type: decoratedSymbol, symbolColor: default, character: f}
+  usePrePhraseSnapshot: true
+  action: {name: remove}
+initialState:
+  documentContents: |2
+        foo
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/scopes/token/ditchFine.yml:12]: A bit hard to read, but there is trailing whitespace here 😅

File: src/test/suite/fixtures/recorded/scopes/token/ditchFine.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  selections:
+    - anchor: {line: 1, character: 0}
+      active: {line: 1, character: 0}
+  marks:
+    default.f:
+      start: {line: 0, character: 4}
+      end: {line: 0, character: 7}
+finalState:
+  documentContents: |+
+
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/scopes/token/ditchFine.yml:22]: ...and no whitespace here

File: src/test/suite/fixtures/recorded/scopes/token/ditchFinePastBat.yml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  marks:
+    default.f:
+      start: {line: 0, character: 4}
+      end: {line: 0, character: 7}
+    default.b:
+      start: {line: 1, character: 4}
+      end: {line: 1, character: 7}
+finalState:
+  documentContents: |2
+        .
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/scopes/token/ditchFinePastBat.yml:33]: This test was broken in your version, because you were looking to see if the end of the line containing the start of the range contained the end of the range, rather than the line containing the end of the range contained the end of the range

File: src/test/suite/fixtures/recorded/scopes/token/ditchFinePastPoint2.yml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        type: primitive
+        mark: {type: decoratedSymbol, symbolColor: default, character: .}
+      excludeAnchor: false
+      excludeActive: false
+  usePrePhraseSnapshot: true
+  action: {name: remove}
+initialState:
+  documentContents: |2
+       foo
+       bar.
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/scopes/token/ditchFinePastPoint2.yml:20]: note the trailing whitespace here

File: src/util/nodeMatchers.ts [hunk 9]
```diff
@@ -74,7 +74,7 @@ export function chainedMatcher(
  * them in reverse, walking up the ancestor chain from `node`.
  * Returns `null` if any finder in the chain returns null.  For example:
  *
- * ancestorChainNodeFinder(0, patternFinder("foo", "bar"), patternFinder("bongo"))
+ * ancestorChainNodeFinder([patternFinder("foo", "bar"), patternFinder("bongo")], 0)
```
Comment (pokey) [inline src/util/nodeMatchers.ts:77]: Noticed this while I was poking around the code to find places where we return a removal range 🤷‍♂️

**Review by pokey — APPROVED** [review]
Ok I made a bunch of changes; have a look and merge if you're happy

File: src/util/nodeMatchers.ts [hunk 10]
```diff
@@ -74,7 +74,7 @@ export function chainedMatcher(
  * them in reverse, walking up the ancestor chain from `node`.
  * Returns `null` if any finder in the chain returns null.  For example:
  *
- * ancestorChainNodeFinder(0, patternFinder("foo", "bar"), patternFinder("bongo"))
+ * ancestorChainNodeMatcher([patternFinder("foo", "bar"), patternFinder("bongo")], 0)
```
Comment (pokey) [inline src/util/nodeMatchers.ts:77]: I noticed this while I was poking around the code to find places where we return a removal range 🤷‍♂️

**Review by pokey — APPROVED** [review]
(see above review; just re-approving to make branch protection happy)

File: src/processTargets/targetUtil/insertionRemovalBehaviors/DelimitedSequenceInsertionRemovalBehavior.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/processTargets/targetUtil/insertionRemovalBehaviors/DelimitedSequenceInsertionRemovalBehavior.ts:14]: This is a regression. Granted in some weird edge cases. Unfortunately we were lacking tests.
in markdown for example
`"# Name "`
`chuck name` will on main remove that trailing white space. Your changes will leave the white space.

File: src/processTargets/targets/ScopeTypeTarget.ts (hunk 4 above)
Comment (AndreasArvidsson) [reply src/processTargets/targets/ScopeTypeTarget.ts:93]: This is a regression. Granted in some weird edge cases. Unfortunately we were lacking tests.
in markdown for example
`"# Name "`
`chuck name` will on main remove that trailing white space. Your changes will leave the white space.

Comment (pokey) [reply src/processTargets/targets/ScopeTypeTarget.ts:93]: As discussed at meet-up, let's keep this removed for now and handle it as part of #855
